### PR TITLE
Core/SAI: Add a warning when a boolean value is outside of [0,1] range

### PR DESCRIPTION
**Pull request by jackpoz**
**Changes proposed:**

- Change SAI fields from uint32 to a new typedef SAIBool (still mapped to uint32) to make it more obvious that it's a boolean value
- Add a warning when a boolean value is outside of [0,1] range. For now no event is skipped, in the future they will be.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:**

None

**Tests performed:**

Tried setting some boolean values to 2 and got a warning
image

**Known issues and TODO list:** (add/remove lines as needed)

- [x] Validate all SAIBool fields
- [ ] Fix the reported issues
- [ ] Skip rows that fail validation

### CODE REVIEW

**Comment by jackpoz** [comment]
@offl this PR spotted these "wrong" values in current SAI. There are some cases with 5000 used as bool (instead of 0 or 1), some with 3516, 653, 2. They most likely have the value in the wrong column and should be checked and fixed, maybe before merging this PR.

>SmartAIMgr: Entry 253 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 2921 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 7790 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 7853 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 15491 SourceType 0 Event 8 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 15491 SourceType 0 Event 9 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 17826 SourceType 0 Event 10 Action 1 uses param e.event.kill.playerOnly of type Boolean with value 5000, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 27199 SourceType 0 Event 1 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 27199 SourceType 0 Event 3 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 28308 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 1 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 2 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 3 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 4 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 5 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 6 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 7 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 8 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 9 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 10 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 11 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 12 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 13 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 29261 SourceType 0 Event 14 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 33522 SourceType 0 Event 0 Action 80 uses param e.action.timedActionList.allowOverride of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 1835100 SourceType 9 Event 3 Action 20 uses param e.action.autoAttack.attack of type Boolean with value 9, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 2628700 SourceType 9 Event 1 Action 53 uses param e.action.wpStart.repeat of type Boolean with value 2, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113500 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113500 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113501 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113501 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113502 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113502 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113503 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113503 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113504 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113504 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113505 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.transport of type Boolean with value 3516, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 3113505 SourceType 9 Event 0 Action 69 uses param e.action.moveToPos.disablePathfinding of type Boolean with value 653, valid values are 0 or 1, skipped.
SmartAIMgr: Entry 12939100 SourceType 9 Event 11 Action 1 uses param e.target.closestFriendly.playerOnly of type Boolean with value 50, valid values are 0 or 1, skipped.

File: src/server/game/AI/SmartScripts/SmartScriptMgr.cpp [hunk 1]
```diff
@@ -32,6 +32,17 @@
 #include "Unit.h"
 #include "WaypointDefines.h"
 
+#define TC_SAI_IS_BOOLEAN_VALID(e, value) \
+{ \
+    if (value > 1) \
+    { \
+        TC_LOG_ERROR("sql.sql", "SmartAIMgr: Entry %d SourceType %u Event %u Action %u uses param %s of type Boolean with value %u, valid values are 0 or 1, skipped.", \
+            e.entryOrGuid, e.GetScriptType(), e.event_id, e.GetActionType(), STRINGIZE(value), value); \
+        /* ToDo: return false once all existing issues have been fixed */ \
```
Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:41]: this is a todo that needs to be completed before merging

**Comment by offl** [comment]
> @offl this PR spotted these "wrong" values in current SAI. There are some cases with 5000 used as bool (instead of 0 or 1), some with 3516, 653, 2. They most likely have the value in the wrong column and should be checked and fixed, maybe before merging this PR.

Not sure if I can do it anytime soon

**Comment by jackpoz** [comment]
I can help research and comment some of the issues, some of these have been bugged for 9 years already so there's no hurry anyway. The checks are supposed to help people that accidentally set a parameter in the wrong column (like npc id in the "run" column)

**Comment by offl** [comment]
Try now

**Comment by jackpoz** [comment]
No more reported errors with latest SAI updates

**Pull request merged**